src/serac/physics/heat_transfer.hpp

```diff
@@ -297,23 +330,42 @@ class HeatTransfer<order, dim, Parameters<parameter_space...>, std::integer_sequ
    * @tparam SourceType The type of the source function
    * @param source_function A source function for a prescribed thermal load
    *
-   * @pre SourceType must have the operator (x, time, temperature, d temperature_dx) defined as the thermal source
+   * @pre source_function must be a object that can be called with the following arguments:
+   *    1. `tensor<T,dim> x` the spatial coordinates for the quadrature point
+   *    2. `double t` the time (note: time will be handled differently in the future)
+   *    3. `T temperature` the current temperature at the quadrature point
+   *    4. `tensor<T,dim>` the spatial gradient of the temperature at the quadrature point
+   *    5. `tuple{value, derivative}`, a variadic list of tuples (each with a values and derivative),
+   *            one tuple for each of the trial spaces specified in the `DependsOn<...>` argument.
+   *
+   * @note The actual types of these arguments passed will be `double`, `tensor<double, ... >` or tuples thereof
+   *    when doing direct evaluation. When differentiating with respect to one of the inputs, its stored
+   *    values will change to `dual` numbers rather than `double`. (e.g. `tensor<double,3>` becomes `tensor<dual<...>,
+   * 3>`)
    */
   template <int... active_parameters, typename SourceType>
   void setSource(DependsOn<active_parameters...>, SourceType source_function)
   {
-    K_functional_->AddDomainIntegral(
-        Dimension<dim>{}, DependsOn<0, active_parameters + 1 ...>{},
-        [source_function, this](auto x, auto temperature, auto... params) {
+    residual_->AddDomainIntegral(
+        Dimension<dim>{}, DependsOn<0, 1, 2, active_parameters + NUM_STATE_VARS...>{},
+        [source_function, this](auto x, auto temperature, auto /* dtemp_dt */, auto shape, auto... params) {
           // Get the value and the gradient from the input tuple
-          auto [u, du_dx] = temperature;
+          auto [u, du_dX] = temperature;
+          auto [p, dp_dX] = shape;
+
+          auto I_plus_dp_dX = I + dp_dX;
+
+          // Note that the current configuration x = X + p, where X is the original reference
+          // configuration and p is the shape displacement. We need the gradient with
+          // respect to the perturbed reference configuration x = X + p for the material model. Therefore, we calculate
+          // du/dx = du/dX * dX/dx = du/dX * (dx/dX)^-1 = du/dX * (I + dp/dX)^-1
 
-          auto flux = serac::zero{};
+          auto du_dx = dot(du_dX, inv(I_plus_dp_dX));
 
-          auto source = -1.0 * source_function(x, time_, u, du_dx, params...);
+          auto source = -1.0 * source_function(x + p, time_, u, du_dx, params...);
```

Comment:
This is reminding me to make patch tests which will check the sign convention of the loads (and more importantly, ensure it doesn't get broken in the future).